apps/pipelines/nodes/nodes.py

```diff
@@ -72,31 +70,118 @@ class CreateReport(LLMResponse):
+        )
+        output = chain.invoke(state["messages"][-1])
+        return output
+
+    def get_participant_data(self, state: PipelineState):
+        session = self.experiment_session(state)
+        return session.get_participant_data()
+
+    @staticmethod
+    def to_json_schema(data: dict):
+        """Converts a dictionary to a JSON schema by first converting it to a Pydantic object and dumping it again.
+        The input should be in the format {"key": "description", "key2": [{"key": "description"}]}
+
+        Nested objects are not supported at the moment
+
+        Input example 1:
+        {"name": "the user's name", "surname": "the user's surname"}
+
+        Input example 2:
+        {"name": "the user's name", "pets": [{"name": "the pet's name": "type": "the type of animal"}]}
+
+        """
+
+        def _create_model_from_data(value_data, model_name: str):
+            pydantic_schema = {}
+            for key, value in value_data.items():
+                if isinstance(value, str):
+                    pydantic_schema[key] = (str | None, Field(description=value))
+                elif isinstance(value, list):
+                    model = _create_model_from_data(value[0], key.capitalize())
+                    pydantic_schema[key] = (list[model], Field(description=f"A list of {key}"))
+            return create_model(model_name, **pydantic_schema)
+
+        Model = _create_model_from_data(data, "CustomModel")
+        schema = Model.model_json_schema()
+        # The schema needs a description in order to comply with function calling APIs
+        schema["description"] = ""
+        return schema
+
+
+class UpdateParticipantMemory(PipelineNode):
+    """A simple component to merge the the input data into the participant's memory data. If key_name is specified
+    the input data will be merged with `key_name` as the key name.
+    """
+
+    __human_name__ = "Update participant memory"
+    key_name: str | None = None
+
+    def _process(self, state: PipelineState) -> PipelineState:
+        extracted_data = state["messages"][-1]
+
+        if not isinstance(extracted_data, dict):
+            # TODO: This should not happen, but until we do input-output type matching, it's always possible
+            input_type = type(extracted_data)
+            raise TypeError(
+                f"UpdateParticipantMemory expected a dictionary as input, but got {input_type} instead.\n"
+                f"Input: {extracted_data}"
+            )
 
-        return RunnableLambda(fn, name=self.__class__.__name__)
+        session = self.experiment_session(state)
+        try:
+            participant_data = ParticipantData.objects.for_experiment(session.experiment).get(
+                participant=session.participant
+            )
+            if self.key_name:
+                participant_data.data[self.key_name] = participant_data.data.get(self.key_name, {}) | extracted_data
```

Comment:
What if the value at `key_name` is a string? Might need a slightly more sophisticated merge method.